homeassistant/components/sensor/filter.py

```diff
@@ -69,6 +75,8 @@
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
     vol.Required(CONF_ENTITY_ID): cv.entity_id,
     vol.Optional(CONF_NAME): cv.string,
+    vol.Optional(CONF_HISTORY_PERIOD): vol.All(cv.time_period,
```

Comment:
I don't think that this should be configurable. We should just fetch based on the maximum window size of the filters.

Reply:
Issue is: some filters have a time window, others have a states window (number of states in the window).

* I can agree when the filter has a time window
* If there is a states window I don't know the period of the sensor to derive the history_period (some sensors aren't even periodic) so my option was to let the user take control.

Reply:
But then should we have at least some sane default ? For example, look for max window size or else set it to 30 minutes?

Reply:
Can I work on the history component and add a:

def get_last_state_changes(hass, number_of_states, entity_id=None):

?

This method would basically query all previous states, order them descending, and limit the returned results to *number_of_states*. The results set would include the datetime of each state.

This method would enable me to solve this issue.

Reply:
Yes.

Reply:
But wouldn't you still want to pass a window size when it's available ?

Reply:
1. Should I open a new PR for the new function or use this PR (since its related) ?
2. Being able to determine the window size, I can't see any use case in which the user would define it's own window. Either he would define a larger window (wastes states) or a smaller window (would still be useful but would not properly restore the filter states)

Reply:
I would not expect users to configure a window directly for this feature, however the user might define a window size for other filters, like time_sma.

This PR is fine.

Reply:
Due to chaining there is only 1 window for all filters in the chain.

The code I will provide will query the window (time/number of state) of all used filters in the filter sensor. Merge all the states (probably through a set()) and initialize all the filters in the chain with the same states, thus reproducing the previous state.